crypto/evp_extra/p_kem.c

```diff
@@ -0,0 +1,339 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <openssl/evp.h>
+
+#include <openssl/err.h>
+#include <openssl/mem.h>
+
+#include "../fipsmodule/evp/internal.h"
+#include "../fipsmodule/delocate.h"
+#include "../kem/internal.h"
+#include "../internal.h"
+#include "internal.h"
+
+typedef struct {
+  const KEM *kem;
+} KEM_PKEY_CTX;
+
+static int pkey_kem_init(EVP_PKEY_CTX *ctx) {
+  KEM_PKEY_CTX *dctx;
+  dctx = OPENSSL_malloc(sizeof(KEM_PKEY_CTX));
+  if (dctx == NULL) {
+    return 0;
+  }
+  OPENSSL_memset(dctx, 0, sizeof(KEM_PKEY_CTX));
+
+  ctx->data = dctx;
+
+  return 1;
+}
+
+static void pkey_kem_cleanup(EVP_PKEY_CTX *ctx) {
+  OPENSSL_free(ctx->data);
+}
+
+static int pkey_kem_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
+  KEM_PKEY_CTX *dctx = ctx->data;
```

Comment:
This variable doesn't server a purpose.  I would just replace the next line with `const KEM *kem = ctx->data->kem`